addons/l10n_tw_edi_ecpay/models/account_move.py

```diff
@@ -0,0 +1,281 @@
+
+        self.l10n_tw_edi_state = "valid" if response_data.get("IIS_Invalid_Status") == "0" else "invalid"
+
+    def _l10n_tw_edi_run_invoice_invalid(self):
+        """
+        Cancelling the e-invoice by calling the Ecpay API and update the invoice information in Odoo
+        """
+        self.ensure_one()
+        if not self.l10n_tw_edi_ecpay_invoice_id:
+            raise UserError(_("You cannot invalidate an invoice that was not sent to Ecpay."))
+        if self.l10n_tw_edi_state == "invalid":
+            raise UserError(_("The invoice: '%(invoice_id)s' has already been invalidated", invoice_id=self.l10n_tw_edi_ecpay_invoice_id))
+
+        json_data = {
+            "MerchantID": self.company_id.l10n_tw_edi_ecpay_merchant_id,
+            "InvoiceNo": self.l10n_tw_edi_ecpay_invoice_id,
+            "InvoiceDate": self.l10n_tw_edi_invoice_create_date.strftime("%Y-%m-%d %H:%M:%S"),
+            "Reason": self.name
+        }
+
+        response_data = EcPayAPI(self.company_id).call_ecpay_api("/Invalid", json_data)
+
+        if response_data.get("RtnCode") != 1:
+            raise UserError(_("Fail to invalidate invoice. Error message: '%(error_message)s'", error_message=response_data.get("RtnMsg")))
+
+        # update the invoice information in Odoo
+        self._l10n_tw_edi_update_ecpay_invoice_info()
+
+    def l10n_tw_edi_issue_allowance(self):
+        """
+        Issuing an allowance by calling the Ecpay API and update the refund invoice information in Odoo
+        Two methods to issue the allowance
+        1. Endpoint: /Allowance
+            General allowance, which requires merchants or sellers to get the agreement from the customer first (not by using ECPay's system)
+            and then to send an API request to ECPay to issue an allowance.
+        2. Endpoint: /AllowanceByCollegiate
+            Sending an API request to ECPay and ECPay will send an e-mail notification with a link to the customer to get his/her agreement
+            ONce the customer clicks the link, an allowance will be issued instantly
+        """
+        self.ensure_one()
+        if not self.l10n_tw_edi_ecpay_invoice_id:
+            raise UserError(_("You cannot refund an invoice that was not sent to Ecpay."))
+
+        json_data = {}
+
+        if self.l10n_tw_edi_refund_agreement_type == "online":
+            if not self.partner_id.email:
+                raise UserError(_("Customer email is needed for notification"))
+            query_param = "/AllowanceByCollegiate"
+            json_data["ReturnURL"] = urljoin(self.env["ir.config_parameter"].sudo().get_param("web.base.url"), f"/invoice/ecpay/agreed_invoice_allowance/{self.id}?access_token={self._portal_ensure_token()}")
+        else:
+            query_param = "/Allowance"
+
+        if self.partner_id.email:
+            json_data["AllowanceNotify"] = "E"
+        elif self.partner_id.phone:
+            json_data["AllowanceNotify"] = "S"
```

Comment:
What if they have both but prefer to get it by phone and not email?
Shouldn't this be something that can be set in the credit note wizard for example?